draft-ietf-httpbis-incremental.md

```diff
@@ -89,38 +90,52 @@ downstream before receiving the complete message.
 
 {::boilerplate bcp14-tagged}
 
-The term Boolean is imported from {{!STRUCTURED-FIELDS=RFC8941}}.
-
 
 # The Incremental Header Field
 
-The Incremental HTTP header field expresses the sender's intent for HTTP
-intermediaries to start forwarding the message downstream before the entire
-message is received.
+The Incremental HTTP header field expresses the sender's preference regarding
+when HTTP intermediaries start forwarding the message downstream.
+
+This header field conveys a Structured Field Token
+({{Section 3.3.4 of STRUCTURED-FIELDS}}).
+
+This specification defines the following three Tokens:
+
+* "always": Requires incremental forwarding.
+* "preferred": Prefers incremental forwarding if feasible.
+* "not-preferred": Allow buffering before forwarding.
 
-This header field has just one valid value of type Boolean: "?1".
+Upon receiving a header section with the Incremental header field, HTTP
+intermediaries respect the sender's preference as follows:
 
-~~~
-Incremental = ?1
-~~~
+When the value is "always", intermediaries SHOULD forward the message
```

Comment:
Nitpicking, but doesn't SHOULD (instead of MUST) make "always" effectively the same as "preferred"?

Reply:
The difference is in how the intermediaries behave when they refuse (or cannot).

In case of `always`, the intermediaries generate errors (because they know that things break by not delivering incremental). In case of `preferred` the intermediaries just buffer the message (because they know that incremental delivery is nothing more than a preference).

Reply:
Right, that's my understanding as well. But doesn't it mean that "always" should be a normative MUST instead of SHOULD?

Reply:
That's fair. Even though the Incremental header is advisory (as HTTP stacks are not required to implement every HTTP extension), for those that understand Incremental it is better to use MUSTs.

Reply:
I agree that SHOULD sounds a bit odd for "always". However, thinking of when it would be violated, it would mainly be a case where incremental delivery doesn't actually make sense (too short to break up into chunks, all available at the same time, not able to send until the whole thing is available). In such a case, the message might not be delivered incrementally, but still isn't an error case.

Reply:
Sounds like we agree that portions of the HTTP message ought to be delivered in a timely manner. Assuming that is the case, I think we should use a MUST that covers the cases that Tommy brought up.

Reply:
This really puts a lie to the whole "always" thing for me.  Because an intermediary might not implement, the value is pretty low value.

On the other hand, a simple Boolean has the benefit of asking for, but not expecting, the non-buffered forwarding.  Obviously, if buffering occurs, the request will go badly, but this is always a risk.

I think that our current approach is to strongly request incremental, but not completely depend on that happening.  That's most healthy for the endpoint requesting that treatment, because they need to think about what might happen if the proxy doesn't understand the request OR the proxy refuses the request.

Reply:
> This really puts a lie to the whole "always" thing for me. Because an intermediary might not implement, the value is pretty low value.

We all agree that intermediaries might not understand the header and buffer the request.

But that point is orthogonal to what clients can expect when they know the intermediaries they are talking to have implemented the extension.

If the debate is between SHOULD vs. MUST, then I can live either ways. But I really think we should assure endpoints that, when talking through intermediaries that implement the extension, the endpoints can expect the they would see incremental delivery or get the request rejected.